fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/indexes/MultidimensionalIndexMaintainer.java

```diff
@@ -311,6 +298,36 @@ public CompletableFuture<Void> deleteWhere(Transaction tr, @Nonnull Tuple prefix
         return super.deleteWhere(tr, prefix);
     }
 
+    /**
+     * Traverse from the root of a key expression of a multidimensional index to the {@link DimensionsKeyExpression}.
+     * @param root the root {@link KeyExpression} of the index definition
+     * @return a {@link DimensionsKeyExpression}
+     */
+    @Nonnull
+    public static DimensionsKeyExpression getDimensionsKeyExpression(@Nonnull final KeyExpression root) {
+        if (root instanceof KeyWithValueExpression) {
+            KeyExpression innerKey = ((KeyWithValueExpression)root).getInnerKey();
+            while (innerKey instanceof ThenKeyExpression) {
+                innerKey = ((ThenKeyExpression)innerKey).getChildren().get(0);
+            }
+            if (innerKey instanceof DimensionsKeyExpression) {
+                return (DimensionsKeyExpression)innerKey;
+            }
+            throw new RecordCoreException("structure of multidimensional index is not supported");
+        }
```

Comment:
This is to support `KeyExpressionWithValue` which is not kept under `DimensionsKeyExpression` which in turn always only separates into prefix, dimensions, suffix (suffix being a key part). `DimensionsKeyExpression` must be the left-most subtree in the key expression to produce keys. You cannot do `concat(x, dimensions(prefix1, prefix2, dim1, dim2, suffix1))`. But potentially you can do `concat(dimensions(prefix1, prefix2, dim1, dim2, suffix1), x)` if `x` is a designated value. I could somehow allow the suffix to encompass values and then completely disallow any `concat` above the `DimensionsKeyExpression`. It, however, turned out to be more complicated and unintuitive compared to what I ended up with.

Reply:
But only one or the other way of combining dimensions and with-value emerges from the building functions / passes validation, correct? And the multiply-nested case doesn't arise from that, either, I think.

If `KeyWithValueExpression` had `keyExpression` and `valueExpression` as the primary state and cached `innerKey`, it might be clearer. (Not that there is any reason to do that.)

Reply:
I think there is a line of code that ensures `KeyWithValueExpression` is the top expression if it is encountered. So yes, the described way is the only way I think that would support the use case and be sensible (and pass validation) wrt `KeyWithValueExpression`. I think I can make the validation stronger to ensure that there is no `ThenKeyExpression` that prefixes the `DimensionsKeyExpression` entirely.

Reply:
I will write more validation code to make sure that the key part of the `KeyWithValue` is exactly matched by the columns owned by the `DimensionsKeyExpression`.

Reply:
Also adding some comments on the expected structure of the key expressions.